### DISCUSSION TITLE
llama.cpp server: How to effectively use cache_prompt parameter

### DISCUSSION THREAD

**Discussion by Mushoz** [ggml-org/llama.cpp#10311] [Q&A]
How are you guys effectively using the cache_prompt parameter? I am using my local LLM with openweb ui and aider, but AFAIK neither has the ability to set the cache_prompt parameter in their requests. So how are you guys effectively enabling the cache with so many clients not supporting this parameter?

Is there some other way to enable caching? The prompt processing would be a lot faster if earlier parts of our conversation could be cached.

**Comment by ggerganov** [comment]
It's up to the clients to support it. On our side, we can add an option to enable it by default for all requests. Maybe it's not a bad idea since I think it's always better to have it enabled.

**Reply by Mushoz** [reply]
Allowing us to enable it globally for all requests would be very welcome!

**Reply by ggerganov** [reply]
Please open a feature request in the issues. It is a good first issue for people to contribute to the project.

**Reply by Mushoz** [reply]
I would be more than happy to open a feature request for this. Just to make sure I will write a feature request with the correct requirements:

1. Should we keep supporting the `cache_prompt` parameter in requests to the server, or should this be removed altogether?
2. Should we simply enable caching for the server and make it non-configurable?
3. Or should we make it configurable as a switch to the server binary? If so, should this default to on or off?

**Reply by vitobotta** [reply]
Is there a workaround for now?

**Reply by ggerganov** [reply]
The feature request should state:

- Change `llama-server` to use `cache_prompt = true` by default
- No need to be able to disable it
- Update docs to explain that the server will cache prompts
- Update clients to not send `cache_prompt` with requests
- Update server to ignore the `cache_prompt` from incoming requests

**Reply by Mushoz** [reply]
@ggerganov is caching ALWAYS desirable though? Some projects send out the same prompt multiple times with non-zero temperature to generate multiple solutions, and then iterate on those. I am not familiar with how caching works, but if the same prompt always returns the same response, then approaching such as those will become impossible.

**Reply by ggerganov** [reply]
Caching the prompt does not affect the generation after that. If you use different seeds then you will get different generations, even if the prompt is cached.

**Reply by steampunque** [reply]
> @ggerganov is caching ALWAYS desirable though? 

The llama.cpp server prompt cache implementation will make generation non-deterministic, meaning you will get different answers for the same submitted prompt.  Therefore if you need deterministic responses (guaranteed to give exact same results for same prompt every time) it will be necessary to turn the prompt cache off.  https://github.com/ggerganov/llama.cpp/pull/7745

**Reply by ggerganov** [reply]
Submitting the same prompt with or without caching will always be deterministic in the single-user case. For multi-user, it's not deterministic with or without prompt caching.

**Reply by ggerganov** [reply]
Correction - I read your https://github.com/ggerganov/llama.cpp/pull/7745#issuecomment-2147939037 and this would indeed lead to non-deterministic results when resubmitting the prompt after it was cached. The solution that separates the last token from the prompt makes sense and we should implement it. It will also be needed for speculative decoding support anyway (#10362)

**Reply by steampunque** [reply]
> Correction - I read your [#7745 (comment)](https://github.com/ggerganov/llama.cpp/pull/7745#issuecomment-2147939037) and this would indeed lead to non-deterministic results when resubmitting the prompt after it was cached. The solution that separates the last token from the prompt makes sense and we should implement it. It will also be needed for speculative decoding support anyway (#10362)

The code I showed in that comment will only work for server config with single slot.  With multiple slots it will be necessary to keep track of the number of tokens added to the accumulating batch on a per slot basis as shown in the partial code block below.  I did not worry about it since I only use single slot config. Any multiple slot config will immediately  bring back non determinism to the results.

```
                       if (/*batch.n_tokens > 1*/ slot.batch_tokens > 1) {
                           slot.start_token = batch.token[batch.n_tokens-1];
                           slot.start_pos = batch.pos[batch.n_tokens-1];
                           slot.have_start_token=true;
                           batch.n_tokens--;
                           slot.batch_tokens--;
```

**Reply by drunnells** [reply]
> Is there a workaround for now?

The OP specifically mentions Open-WebUI. I've made a pull request to add cache_prompt to the completions call: https://github.com/open-webui/open-webui/pull/7237

Or in the mean time you can manually edit site-packages/open_webui/apps/openai/main.py to add:

`    payload["cache_prompt"] = True`

Right before:    
```
    # Convert the modified body back to JSON
    payload = json.dumps(payload)
```
